<a href="https://colab.research.google.com/github/MariaReyesMedina/Homework3-DACSS690C/blob/main/Homework3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
!ls '/content/drive/MyDrive/Colab Notebooks/'

 cars.csv
'Copia de Copia 2 de DACSS 603 Tutorial 6_ANOVA & Chi-Square.ipynb'
'Copia de DACSS 603 Tutorial2_Probability_Distributions.ipynb'
'Copia de Homework3.ipynb'
 Employee_Salary_Dataset.csv
'Este si Maria de DACSS 603 Tutorial 7_Simple_Linear_Regression.ipynb'
 Homework3.ipynb
'Maria Copia de DACSS 603 Tutorial 10_Transformations.ipynb'
'Maria Copia de DACSS 603 Tutorial 8__Multiple_Regression_I.ipynb'
'Maria Copia de DACSS 603 Tutorial 9_Multiple Regression II.ipynb'
'Maria DACSS 603 Tutorial1_Descriptive_Statistics.ipynb'
'Maria DACSS 603 Tutorial 3_CentralLimitTheorem.ipynb'
'Maria DACSS 603 Tutorial 4_Confidence_Intervals.ipynb'
'Maria DACSS 603 Tutorial 5_Hypothesis_Testing.ipynb'
'MariaReyes Homework1 DACSS690C.ipynb'
'Maria Reyes Tutorial 10_DACSS601: Visualizing Time & Space'
'Maria Reyes Tutorial 10_DACSS601: Visualizing Time & Space (1)'
'Maria Reyes Tutorial 1_DACSS601.ipynb'
'Maria Reyes Tutorial 2_DACSS601: Basic Statistics'
'Maria Reyes Tutorial 3_DACSS601: Tidyver

In [3]:
!jupyter nbconvert --to html '/content/drive/MyDrive/Colab Notebooks/Homework3.ipynb'

[NbConvertApp] Converting notebook /content/drive/MyDrive/Colab Notebooks/Homework3.ipynb to html
[NbConvertApp] Writing 366052 bytes to /content/drive/MyDrive/Colab Notebooks/Homework3.html


Introduction to Optimization for Decision Making
Maria Reyes
DACSS690C Summer 2026

In [ ]:
!wget https://github.com/MariaReyesMedina/Homework3-DACSS690C/raw/main/README.md

--2026-08-12 23:17:46--  https://github.com/MariaReyesMedina/Homework3-DACSS690C/raw/main/README.md
Resolving github.com (github.com)... 140.82.113.3
Connecting to github.com (github.com)|140.82.113.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/MariaReyesMedina/Homework3-DACSS690C/main/README.md [following]
--2026-08-12 23:17:46--  https://raw.githubusercontent.com/MariaReyesMedina/Homework3-DACSS690C/main/README.md
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 22 [text/plain]
Saving to: ‘README.md.1’

README.md.1         100%[===================>]      22  --.-KB/s    in 0s      

2026-08-12 23:17:46 (414 KB/s) - ‘README.md.1’ saved [22/22]



In [ ]:
linkAHP='https://github.com/MariaReyesMedina/Homework3-DACSS690C/raw/main/ahp%20school%20choice.xlsx'

In [ ]:
pairwise_learning=pd.read_excel(linkAHP,sheet_name='learning', index_col=0)
pairwise_learning

,a,b,c
a,1,NaN,NaN
b,3,1.0,3.0
c,2,NaN,1.0


In [ ]:
pairwise_frinds=pd.read_excel(linkAHP,sheet_name='frinds', index_col=0)
pairwise_frinds

,a,b,c
a,1,1,1
b,1,1,1
c,1,1,1


In [ ]:
pairwise_school_life=pd.read_excel(linkAHP,sheet_name='school life', index_col=0)
pairwise_school_life

,a,b,c
a,1.0,5,1.0
b,NaN,1,NaN
c,1.0,5,1.0


In [ ]:
pairwise_vocational_trng=pd.read_excel(linkAHP,sheet_name='vocational trng', index_col=0)
pairwise_vocational_trng

,a,b,c
a,1.0,9,7.0
b,NaN,1,NaN
c,NaN,5,1.0


In [ ]:
pairwise_collage_prep=pd.read_excel(linkAHP,sheet_name='collage prep', index_col=0)
pairwise_collage_prep


,a,b,c
a,1,NaN,1
b,2,1.0,2
c,1,NaN,1


In [ ]:
pairwise_music_classes=pd.read_excel(linkAHP,sheet_name='music classes', index_col=0)
pairwise_music_classes


,a,b,c
a,1.0,6,4.0
b,NaN,1,NaN
c,NaN,3,1.0


In [ ]:
pairwise_criteria=pd.read_excel(linkAHP,sheet_name='criteria', index_col=0)
pairwise_criteria

,learning,frinds,school life,vocational trng,collage prep,music classes
learning,1.0,4.0,3,1.0,3.0,4.0
frinds,NaN,1.0,7,3.0,NaN,1.0
school life,NaN,NaN,1,NaN,NaN,NaN
vocational trng,1.0,NaN,5,1.0,1.0,NaN
collage prep,NaN,5.0,5,1.0,1.0,3.0
music classes,NaN,1.0,6,3.0,NaN,1.0


All the comparissons:

In [ ]:
import networkx as nx

G_learning = nx.from_pandas_adjacency(pairwise_learning, # just change this pairwase
                                 create_using=nx.DiGraph())

# pairwise
G_learning.edges(data=True)


import numpy as np
learning_comparisons ={(e[0],e[1]):e[2]['weight'] for e in G_learning.edges(data=True) if np.isfinite(e[2]['weight'])}

# ... and the remaining comparissons:
G_frinds = nx.from_pandas_adjacency(pairwise_frinds,create_using=nx.DiGraph())
frinds_comparisons={(e[0],e[1]):e[2]['weight'] for e in G_frinds.edges(data=True) if np.isfinite(e[2]['weight'])}

G_school_life = nx.from_pandas_adjacency(pairwise_school_life,create_using=nx.DiGraph())
school_life_comparisons={(e[0],e[1]):e[2]['weight'] for e in G_school_life.edges(data=True) if np.isfinite(e[2]['weight'])}

G_vocational_trng = nx.from_pandas_adjacency(pairwise_vocational_trng,create_using=nx.DiGraph())
vocational_trng_comparisons={(e[0],e[1]):e[2]['weight'] for e in G_vocational_trng.edges(data=True) if np.isfinite(e[2]['weight'])}

G_collage_prep = nx.from_pandas_adjacency(pairwise_collage_prep,create_using=nx.DiGraph())
collage_prep_comparisons={(e[0],e[1]):e[2]['weight'] for e in G_collage_prep.edges(data=True) if np.isfinite(e[2]['weight'])}

G_music_classes = nx.from_pandas_adjacency(pairwise_music_classes,create_using=nx.DiGraph())
music_classes_comparisons={(e[0],e[1]):e[2]['weight'] for e in G_music_classes.edges(data=True) if np.isfinite(e[2]['weight'])}



In [ ]:
# take a look
[learning_comparisons, frinds_comparisons,vocational_trng_comparisons,collage_prep_comparisons,music_classes_comparisons]

[{('a', 'a'): 1.0,
  ('b', 'a'): 3.0,
  ('b', 'b'): 1.0,
  ('b', 'c'): 3.0,
  ('c', 'a'): 2.0,
  ('c', 'c'): 1.0},
 {('a', 'a'): 1,
  ('a', 'b'): 1,
  ('a', 'c'): 1,
  ('b', 'a'): 1,
  ('b', 'b'): 1,
  ('b', 'c'): 1,
  ('c', 'a'): 1,
  ('c', 'b'): 1,
  ('c', 'c'): 1},
 {('a', 'a'): 1.0,
  ('a', 'b'): 9.0,
  ('a', 'c'): 7.0,
  ('b', 'b'): 1.0,
  ('c', 'b'): 5.0,
  ('c', 'c'): 1.0},
 {('a', 'a'): 1.0,
  ('a', 'c'): 1.0,
  ('b', 'a'): 2.0,
  ('b', 'b'): 1.0,
  ('b', 'c'): 2.0,
  ('c', 'a'): 1.0,
  ('c', 'c'): 1.0},
 {('a', 'a'): 1.0,
  ('a', 'b'): 6.0,
  ('a', 'c'): 4.0,
  ('b', 'b'): 1.0,
  ('c', 'b'): 3.0,
  ('c', 'c'): 1.0}]

In [ ]:
# now the criteria

G_CRIT = nx.from_pandas_adjacency(pairwise_criteria,create_using=nx.DiGraph())
criteria_comparisons ={(e[0],e[1]):e[2]['weight'] for e in G_CRIT.edges(data=True) if np.isfinite(e[2]['weight'])}
criteria_comparisons

{('learning', 'learning'): 1.0,
 ('learning', 'frinds'): 4.0,
 ('learning', 'school life '): 3.0,
 ('learning', 'vocational trng'): 1.0,
 ('learning', 'collage prep'): 3.0,
 ('learning', 'music classes'): 4.0,
 ('frinds', 'frinds'): 1.0,
 ('frinds', 'school life '): 7.0,
 ('frinds', 'vocational trng'): 3.0,
 ('frinds', 'music classes'): 1.0,
 ('school life ', 'school life '): 1.0,
 ('vocational trng', 'learning'): 1.0,
 ('vocational trng', 'school life '): 5.0,
 ('vocational trng', 'vocational trng'): 1.0,
 ('vocational trng', 'collage prep'): 1.0,
 ('collage prep', 'frinds'): 5.0,
 ('collage prep', 'school life '): 5.0,
 ('collage prep', 'vocational trng'): 1.0,
 ('collage prep', 'collage prep'): 1.0,
 ('collage prep', 'music classes'): 3.0,
 ('music classes', 'frinds'): 1.0,
 ('music classes', 'school life '): 6.0,
 ('music classes', 'vocational trng'): 3.0,
 ('music classes', 'music classes'): 1.0}

Now AHP

In [ ]:
!pip install AHPy

In [ ]:
import ahpy

learning = ahpy.Compare('learning', learning_comparisons, random_index='saaty')
frinds = ahpy.Compare('frinds', frinds_comparisons, random_index='saaty')
school_life = ahpy.Compare('school life', school_life_comparisons, random_index='saaty')
vocational_trng = ahpy.Compare('vocational trng', vocational_trng_comparisons, random_index='saaty')
collage_prep = ahpy.Compare('collage prep', collage_prep_comparisons, random_index='saaty')
music_classes = ahpy.Compare('music classes', music_classes_comparisons, random_index='saaty')
criteria = ahpy.Compare('criteria', criteria_comparisons, random_index='saaty')

Create hierarchy:


In [ ]:
criteria.add_children([learning, frinds, school_life, vocational_trng, collage_prep, music_classes])

Which criterion was more valuable: Learning

In [ ]:
weigths=criteria.global_weights
pd.Series(weigths).to_frame(name='Value')

,Value
learning,0.3208
collage prep,0.2374
frinds,0.1395
music classes,0.1391
vocational trng,0.1285
school life,0.0348


Result:

In [ ]:
priorities=criteria.target_weights
pd.Series(priorities).to_frame(name='Value')

,Value
b,0.3753
a,0.3516
c,0.2384


Assess consistency

In [ ]:
assesment=[(val.name,val.consistency_ratio) for val in [learning, frinds, school_life, vocational_trng, collage_prep, music_classes]]
pd.DataFrame(assesment,columns=['Criterion', 'Consistency Ratio'])

,Criterion,Consistency Ratio
0,learning,0.0516
1,frinds,0.0000
2,school life,0.0000
3,vocational trng,0.2005
4,collage prep,0.0000
5,music classes,0.0516


The vocational trng criterion has a consistency ratio of 0.20, which is above the 0.10 threshold. This indicates a relatively high level of inconsistency in the pairwise judgments, so these comparisons should be reviewed.